### PR TITLE
Redirect to previous url after login

### LINKED ISSUE #16588: Redirect user to the previous URL after login
Redirect to previous URL after login

Support: https://support.abp.io/QA/Questions/5051/How-can-I-redirect-the-user-to-the-previous-URL-after-login

### PR DESCRIPTION
**Pull request by masum-ulu**
Resolves #16588 

## Code flow
redirect-code-flow

## Password flow
redirect-password-flow

### CODE REVIEW

File: npm/ng-packs/packages/oauth/src/lib/strategies/auth-code-flow-strategy.ts [hunk 1]
```diff
@@ -14,7 +14,10 @@ export class AuthCodeFlowStrategy extends AuthFlowStrategy {
   }
 
   navigateToLogin(queryParams?: Params) {
-    this.oAuthService.initCodeFlow('', this.getCultureParams(queryParams));
+    this.oAuthService.initCodeFlow(
```
Comment (mahmut-gundogdu) [inline npm/ng-packs/packages/oauth/src/lib/strategies/auth-code-flow-strategy.ts]: Can ve move this functions to variable ?

File: npm/ng-packs/packages/oauth/src/lib/strategies/auth-flow-strategy.ts [hunk 2]
```diff
@@ -80,6 +84,28 @@ export abstract class AuthFlowStrategy {
       .catch(this.catchError);
   }
 
+  protected navigateToPreviousUrl() {
+    this.oAuthService.events
```
Comment (mahmut-gundogdu) [inline npm/ng-packs/packages/oauth/src/lib/strategies/auth-flow-strategy.ts]: it should be disposed.

**Review by mahmut-gundogdu — CHANGES_REQUESTED** [review]

File: npm/ng-packs/packages/oauth/src/lib/strategies/auth-flow-strategy.ts (hunk 2 above)
Comment (masum-ulu) [reply npm/ng-packs/packages/oauth/src/lib/strategies/auth-flow-strategy.ts]: I'm not sure how can I dispose this method. Other examples seems like same

https://github.com/abpframework/abp/blob/419c2f2af451691ad15b6b4d834c3663b141288c/npm/ng-packs/packages/oauth/src/lib/strategies/auth-flow-strategy.ts#L69

# 

I just improved with condition

image

File: npm/ng-packs/packages/oauth/src/lib/strategies/auth-flow-strategy.ts [hunk 3]
```diff
@@ -80,6 +81,31 @@ export abstract class AuthFlowStrategy {
       .catch(this.catchError);
   }
 
+  protected navigateToPreviousUrl(): void {
+    const { responseType } = this.oAuthConfig;
+    if (responseType === 'code') {
+      this.oAuthService.events
+        .pipe(
```
Comment (mahmut-gundogdu) [inline npm/ng-packs/packages/oauth/src/lib/strategies/auth-flow-strategy.ts:88]: We may use .take(1) or something. it cause memory leak

File: npm/ng-packs/packages/oauth/src/lib/strategies/auth-flow-strategy.ts [hunk 4]
```diff
@@ -80,6 +81,31 @@ export abstract class AuthFlowStrategy {
...
+            const redirect_uri = decodeURIComponent(this.oAuthService.state);
+
+            if (redirect_uri && redirect_uri !== '/') {
+              return redirect_uri;
+            }
+            return '/';
+          }),
+          switchMap(redirectUri =>
+            this.configState.getOne$('currentUser').pipe(
+              filter(user => Boolean(user?.isAuthenticated)),
```
Comment (mahmut-gundogdu) [inline npm/ng-packs/packages/oauth/src/lib/strategies/auth-flow-strategy.ts]: Boolean(user?.isAuthenticated) is ok but you can use  !!user?.isAuthenticated
double bang (!) . it. just a remider . it is ok

**Review by mahmut-gundogdu — CHANGES_REQUESTED** [review]

File: npm/ng-packs/packages/oauth/src/lib/strategies/auth-flow-strategy.ts (hunk 2 above)
Comment (mahmut-gundogdu) [reply npm/ng-packs/packages/oauth/src/lib/strategies/auth-flow-strategy.ts]: https://rxjs.dev/api/operators/take
.take(1) before the subscribe.  it turn to promise-like obserable.

**Review by mahmut-gundogdu — APPROVED** [review]

**Pull request merged**